### PR TITLE
 Verify source spec at indexing time (#3827)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
Connected issue: #3827 

### Author

- [x] Target branch is `develop`
- [x] Name of source branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references the connected issue
- [x] PR title matches<sup>1</sup> title of connected issue         <sub>or comment in PR explains why they're different</sub>
- [x] Title of at least one commit references connected issue
- [x] PR is connected to issue via Zenhub 
- [x] PR description links to connected issue
- [x] Added `partial` label to PR                                    <sub>or this PR completely resolves the connected issue</sub>

<sup>1</sup> when the issue title describes a problem, the PR title is `Fix: ` followed by the issue title   

### Author (reindex)

- [x] Added `r` tag to commit title                                 <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                                   <sub>or this PR does not require reindexing</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain               <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR                        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR                             <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst          <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                                 <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                                   <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks           <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title                            <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfix connected to the issue          <sub>or there is no temporary hotfix for the connected issue on the `prod` branch</sub>

### Author (requirements)

- [x] Ran `make requirements_update`                                <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title                                 <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR                                      <sub>or this PR does not touch requirements*.txt</sub>

### Author (rebasing, integration test)

- [x] `make integration_test` passes in personal deployment         <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

### Peer reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists.

### Peer reviewer (after approval)

- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *Author (requirements)* and *Author (rebasing, integration test)* 
checklists. Update the `N reviews` label. 

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issue as `demo` or `no demo`
- [x] Commented on connected issue about demo expectations          <sub>or labelled connected issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear                      <sub>or connected issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Branch pushed to GitLab and added `sandbox` label             <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                                       <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox`                     <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [x] Started reindex in `sandbox`                                  <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                             <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved connected issue to Merged column
- [x] Pushed merge commit to GitHub

### Operator (after pushing the merge commit)

- [ ] Shortened the PR chain                                        <sub>or this PR is not labeled `base`</sub>
- [ ] Pushed merge commit to GitLab                                 <sub>or merge commit can be pushed later, with another PR</sub>
- [ ] Deleted PR branch from GitHub and GitLab
- [ ] Build passes on GitLab

### Operator (reindex) 

- [ ] Deleted unreferenced indices in `dev`                         <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices</sub> 
- [ ] Started reindex in `dev`                                      <sub>or this PR does not require reindexing</sub>
- [ ] Checked for and triaged indexing failures                     <sub>or this PR does not require reindexing</sub>
- [ ] Emptied fail queues in target deployment                      <sub>or this PR does not require reindexing</sub>

### Operator

- [ ] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Review by danielsotirhos — APPROVED** [review]
Approved.

File: src/azul/plugins/repository/tdr/__init__.py [hunk 1]
```diff
@@ -160,7 +161,10 @@ def dangling_inputs(self) -> Entities:
 
 
 class TDRSourceRef(SourceRef[TDRSourceSpec, 'TDRSourceRef']):
-    pass
+
+    @property
+    def typed_id(self) -> TypedSourceId:
+        return TypedSourceId(id=self.id, type=self.spec.type_name)
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr/__init__.py:167]: If the TDRSourceRef already has the source ID and the source type, why do we need a separate class? Couldn't we pass the TDRSourceRef instance to the methods that currently require an TypedSourceId?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/plugins/repository/tdr/__init__.py (hunk 1 above)
Comment (noah-aviel-dove) [reply src/azul/plugins/repository/tdr/__init__.py:167]: Importing `TDRSourceRef` into `terra.py` would create a circular dependency.

Comment (noah-aviel-dove) [reply src/azul/plugins/repository/tdr/__init__.py]: Decided at PL to move/rename `TDRSourceRef` to avoid this

**Comment by noah-aviel-dove** [comment]
Re-requesting peer review due to staleness.

**Review by danielsotirhos — APPROVED** [review]
Approved.

File: src/azul/plugins/repository/tdr/__init__.py [hunk 2]
```diff
@@ -240,6 +236,11 @@ def tdr(self):
     def _tdr(cls):
         return TDRClient.for_indexer()
 
+    def verify_source(self, ref: TDRSourceRef) -> None:
+        actual_name = self.tdr.lookup_source_name(ref)
+        require(ref.spec.name == actual_name,
+                'Source name changed unexpectedly', ref, actual_name)
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr/__init__.py:242]: I think we should move the implementation to the client.

```suggestion
        self.tdr.verify_source((ref)
```

and inline `lookup_source_name`.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**